lvgl/src/display.rs

```diff
@@ -0,0 +1,359 @@
+use crate::functions::CoreError;
+use crate::{disp_drv_register, disp_get_default, get_str_act};
+use crate::{Box, RunOnce};
+use crate::{Color, Obj};
+use core::cell::RefCell;
+use core::marker::PhantomData;
+use core::mem::MaybeUninit;
+use core::ptr::NonNull;
+use core::{ptr, result};
+use embedded_graphics::drawable;
+use embedded_graphics::prelude::*;
+use parking_lot::const_mutex;
+use parking_lot::Mutex;
+
+#[cfg(feature = "alloc")]
+use alloc::sync::Arc;
+
+pub const DISP_HOR_RES: usize = lvgl_sys::LV_HOR_RES_MAX as usize;
+pub const DISP_VER_RES: usize = lvgl_sys::LV_VER_RES_MAX as usize;
+
+#[derive(Debug, Copy, Clone, Eq, PartialEq)]
+pub enum DisplayError {
+    NotAvailable,
+    FailedToRegister,
+    NotRegistered,
+}
+
+type Result<T> = result::Result<T, DisplayError>;
+
+#[cfg(feature = "alloc")]
+pub type SharedNativeDisplay<T> = Arc<Mutex<T>>;
+
+pub struct Display {
+    pub(crate) disp: NonNull<lvgl_sys::lv_disp_t>,
+}
+
+impl Display {
+    pub(crate) fn from_raw(disp: NonNull<lvgl_sys::lv_disp_t>) -> Self {
+        Self { disp }
+    }
+
+    pub fn register<T, C, const N: usize>(
+        draw_buffer: &'static DrawBuffer<N>,
+        native_display: T,
+    ) -> Result<Self>
+    where
+        T: DrawTarget<C>,
+        C: PixelColor + From<Color>,
+    {
+        let mut display_diver = DisplayDriver::new(draw_buffer, native_display)?;
+        Ok(disp_drv_register(&mut display_diver)?)
+    }
+
+    #[cfg(feature = "alloc")]
+    pub fn register_shared<T, C, const N: usize>(
+        draw_buffer: &'static DrawBuffer<N>,
+        shared_native_display: &SharedNativeDisplay<T>,
+    ) -> Result<Self>
+    where
+        T: DrawTarget<C>,
+        C: PixelColor + From<Color>,
+    {
+        let mut display_diver =
+            DisplayDriver::new_shared(draw_buffer, Arc::clone(shared_native_display))?;
+        Ok(disp_drv_register(&mut display_diver)?)
+    }
+
+    pub fn get_scr_act(&self) -> Result<Obj> {
+        Ok(get_str_act(Some(&self))?)
+    }
+}
+
+impl Default for Display {
+    fn default() -> Self {
+        disp_get_default().expect("LVGL must be INITIALIZED")
+    }
+}
+
+#[derive(Copy, Clone)]
+pub struct DefaultDisplay {}
+
+impl DefaultDisplay {
+    /// Gets the screen active of the default display.
+    pub fn get_scr_act() -> Result<Obj> {
+        Ok(get_str_act(None)?)
+    }
+}
+
+pub struct DrawBuffer<const N: usize> {
+    initialized: RunOnce,
+    refresh_buffer: Mutex<RefCell<heapless::Vec<lvgl_sys::lv_color_t, N>>>,
```

Comment:
I use a `Mutex` here, to satisfy the Rust guarantees. But maybe it is not necessary... some food for thought.